lib/spacesuit/cors_middleware.ex

```diff
@@ -0,0 +1,166 @@
+defmodule Spacesuit.CorsMiddleware do
+  require Logger
+  use Elixometer
+
+  @http_server            Application.get_env(:spacesuit, :http_server)
+  @cors_config            Application.get_env(:spacesuit, :cors_config)
+  @supported_http_methods [:GET, :POST, :PUT, :PATCH, :DELETE, :HEAD, :OPTIONS]
+
+  @timed(key: "timed.corsMiddleware-handle", units: :millisecond)
+  def execute(req, env) do
+    case {req[:headers]["origin"], req[:method]} do
+      {nil, _} ->
+        # http://www.w3.org/TR/cors/#resource-requests
+        # § 6.1.1
+        # If the Origin header is not present terminate this set of steps.
+        Logger.debug "No origin header on request"
+        {:ok, req, env}
+
+      {origin, method} ->
+        if is_nil(origin) or !is_valid_origin(origin) do
```

Comment:
You can clean up some of these ifs and various functions with some pattern matching like:

```elixir
def serve_from_origin?(origin) do
  case String.split(origin || "", "%") do
    # It was nil or empty string
    [""] -> false
    # It contained a %
    [_, _] -> false
    # String was ok, is the scheme ok?
    [potentially_valid] -> !is_nil(URI.parse(potentially_valid).scheme)
  end
end
```
Note that you can turn a `nil` into a string by ORing it with empty string.

Using that `serve_from_origin?()` method you could also join the lower `if` together, e.g.:
```elixir
if serve_from_origin?(origin) || @cors_config[:serve_forbidden_origins] do
  {:ok, req, env}
end```

Etc.